### PR TITLE
Add support for AMB contracts

### PR DESCRIPTION
**Pull request by patitonar**
This PR includes:
- Oracle support for AMB
- Monitor support for AMB
- Oracle-e2e updated with AMB tests
- Contract submodule updated to a commit that includes AMB work. Once the set of contracts is released, we should update the submodule to the new version. 

monitor-e2e should be updated to include tests for AMB. We can create an issue to later address this.

### CODE REVIEW

File: oracle/src/events/processAMBCollectedSignatures/index.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          eventTransactionHash: colSignature.transactionHash
+        })
+
+        if (authorityResponsibleForRelay === web3Home.utils.toChecksumAddress(config.validatorAddress)) {
+          logger.info(`Processing CollectedSignatures ${colSignature.transactionHash}`)
+          const message = await homeBridge.methods.message(messageHash).call()
+
+          const requiredSignatures = []
+          requiredSignatures.length = NumberOfCollectedSignatures
+          requiredSignatures.fill(0)
```
Comment (rzadp) [inline oracle/src/events/processAMBCollectedSignatures/index.js]: Could be replaced with:
```
const requiredSignatures = (new Array(NumberOfCollectedSignatures)).fill(0)
```

**Comment by patitonar** [comment]
@akolotov Removed everything related to `defrayal` mode

File: oracle/src/events/processAMBCollectedSignatures/index.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+          const gasPriceOptions = generateGasPriceOptions({ dataType, gasPrice, gasPriceSpeed })
+          txToSend.push({
+            data,
+            gasEstimate,
+            transactionReference: colSignature.transactionHash,
+            to: config.foreignBridgeAddress,
+            gasPriceOptions
+          })
+        } else {
+          logger.info(`Validator not responsible for relaying CollectedSignatures ${colSignature.transactionHash}`)
```
[the remark this answers is not in the document]
Comment (patitonar) [reply oracle/src/events/processAMBCollectedSignatures/index.js]: Thanks! Please check https://github.com/poanetwork/tokenbridge/pull/199/commits/0aa3b2074598141ccc044c5912ab23c8c85be9cd

File: deployment/roles/oracle/templates/.env.j2 [hunk 3]
```diff
@@ -9,7 +9,9 @@ HOME_POLLING_INTERVAL={{ HOME_POLLING_INTERVAL }}
 ## Foreign contract
 FOREIGN_RPC_URL={{ FOREIGN_RPC_URL }}
 FOREIGN_BRIDGE_ADDRESS={{ FOREIGN_BRIDGE_ADDRESS }}
+{% if ERC20_TOKEN_ADDRESS | default('') != '' %}
```
Comment (akolotov) [inline deployment/roles/oracle/templates/.env.j2]: it should disappear after merge

File: monitor/alerts.js [hunk 4]
```diff
@@ -13,10 +15,17 @@ const foreignProvider = new Web3.providers.HttpProvider(FOREIGN_RPC_URL)
...
-  const { foreignDeposits, homeDeposits, homeWithdrawals, foreignWithdrawals } = await eventsInfo()
-
-  const xSignatures = foreignDeposits.filter(findDifferences(homeDeposits))
-  const xAffirmations = homeWithdrawals.filter(findDifferences(foreignWithdrawals))
+  const { foreignDeposits, homeDeposits, homeWithdrawals, foreignWithdrawals, bridgeMode } = await eventsInfo()
+
+  let xSignatures
+  let xAffirmations
+  if (bridgeMode === BRIDGE_MODES.ARBITRARY_MESSAGE) {
+    xSignatures = foreignDeposits.filter(processedMsgNotDelivered(homeDeposits))
```
Comment (akolotov) [inline monitor/alerts.js]: please consider either to change names `foreignDeposits`, `homeDeposits`, `homeWithdrawals` and `foreignWithdrawals` to more common (e.g. `foreignToHomeRequests`, `homeToForeignRequests`, `foreignToHomeConfirmations` and `homeToForeignConfirmations`) under this PR or create another issue to address it later.

File: monitor/eventsStats.js [hunk 5]
```diff
@@ -57,25 +58,45 @@ function compareTransferForeign(home) {
 }
 
 async function main() {
-  const { foreignDeposits, homeDeposits, homeWithdrawals, foreignWithdrawals, isExternalErc20 } = await eventsInfo()
+  const {
+    foreignDeposits,
```
Comment (akolotov) [inline monitor/eventsStats.js]: please see comment for changes in `alert.js`

File: monitor/getShortEventStats.js [hunk 6]
```diff
@@ -1,19 +1,35 @@
 require('dotenv').config()
 const eventsInfo = require('./utils/events')
+const { BRIDGE_MODES } = require('../commons')
 
 async function main(bridgeMode) {
   const { foreignDeposits, homeDeposits, homeWithdrawals, foreignWithdrawals } = await eventsInfo(bridgeMode)
```
Comment (akolotov) [inline monitor/getShortEventStats.js]: please see comment for changes in `alert.js`

File: monitor/getShortEventStats.js [hunk 7]
```diff
@@ -1,19 +1,35 @@
...
-    home: {
-      deposits: homeDeposits.length,
-      withdrawals: homeWithdrawals.length
-    },
-    foreign: {
-      deposits: foreignDeposits.length,
-      withdrawals: foreignWithdrawals.length
+  if (bridgeMode === BRIDGE_MODES.ARBITRARY_MESSAGE) {
+    return {
+      deliveryDiff: homeDeposits.length - foreignDeposits.length,
```
Comment (akolotov) [inline monitor/getShortEventStats.js]: for AMB it is necessary to change `checkWorkers.js` as so it outputs:
```
{"home":{"toForeign":1513,"fromForeign":1651},"foreign":{"fromHome":1513,"toHome":1655}, "lastChecked":1568748063,"fromHomeToForeignDiff":0,"fromForeignToHomeDiff":-4,"timeDiff":69}
```

File: monitor/getBalances.js [hunk 8]
```diff
@@ -107,6 +107,22 @@ async function main(bridgeMode) {
       balanceDiff: Number(Web3Utils.fromWei(diff)),
       lastChecked: Math.floor(Date.now() / 1000)
     }
+  } else if (bridgeMode === BRIDGE_MODES.ARBITRARY_MESSAGE) {
```
[the remark this answers is not in the document]
Comment (patitonar) [reply monitor/getBalances.js]: Done https://github.com/poanetwork/tokenbridge/pull/199/commits/ef447b7b2b2a01ab25155aa55d9473fdbdb32df5

File: deployment/roles/oracle/templates/.env.j2 (hunk 3 above)
Comment (patitonar) [reply deployment/roles/oracle/templates/.env.j2]: Yes, it disappeared

File: monitor/alerts.js (hunk 4 above)
Comment (patitonar) [reply monitor/alerts.js]: Thanks for suggestion, fixed https://github.com/poanetwork/tokenbridge/pull/199/commits/4370fd25b4f61c3920f5a4ba63d42afc3045d24e

File: monitor/getShortEventStats.js (hunk 7 above)
Comment (patitonar) [reply monitor/getShortEventStats.js]: Updated the output. Please check https://github.com/poanetwork/tokenbridge/pull/199/commits/9c87c239daa852a311e89545701a2b95e01c6e98

**Pull request merged**